handlers/anteHandler.go

```diff
@@ -16,116 +16,106 @@ import (
 	return func(ctx sdk.Context, tx sdk.Tx, simulate bool) (newCtx sdk.Context, res sdk.Result, abort bool) {
 		msg := tx.GetMsgs()[0] // tx should only have one msg
 		switch mtype := msg.Type(); mtype {
 		case "include_deposit":
 			depositMsg := msg.(msgs.IncludeDepositMsg)
-			return includeDepositAnteHandler(ctx, utxoStore, depositMsg, client)
+			return includeDepositAnteHandler(ctx, outputStore, blockStore, depositMsg, client)
 		case "spend_utxo":
 			spendMsg := msg.(msgs.SpendMsg)
-			return spendMsgAnteHandler(ctx, spendMsg, utxoStore, plasmaStore, client)
+			return spendMsgAnteHandler(ctx, spendMsg, outputStore, blockStore, client)
 		default:
-			return ctx, msgs.ErrInvalidTransaction(DefaultCodespace, "Msg is not of type SpendMsg or IncludeDepositMsg").Result(), true
+			return ctx, ErrInvalidTransaction(DefaultCodespace, "msg is not of type SpendMsg or IncludeDepositMsg").Result(), true
 		}
 	}
 }
 
-func spendMsgAnteHandler(ctx sdk.Context, spendMsg msgs.SpendMsg, utxoStore store.UTXOStore, plasmaStore store.PlasmaStore, client plasmaConn) (newCtx sdk.Context, res sdk.Result, abort bool) {
+func spendMsgAnteHandler(ctx sdk.Context, spendMsg msgs.SpendMsg, outputStore store.OutputStore, blockStore store.BlockStore, client plasmaConn) (newCtx sdk.Context, res sdk.Result, abort bool) {
 	var totalInputAmt, totalOutputAmt *big.Int
+	totalInputAmt = big.NewInt(0)
+	totalOutputAmt = big.NewInt(0)
 
 	// attempt to recover signers
 	signers := spendMsg.GetSigners()
 	if len(signers) == 0 {
-		return ctx, msgs.ErrInvalidTransaction(DefaultCodespace, "failed recovering signers").Result(), true
+		return ctx, ErrInvalidTransaction(DefaultCodespace, "failed recovering signers").Result(), true
 	}
 
-	/* validate the first input */
-	amt, res := validateInput(ctx, spendMsg.Input0, common.BytesToAddress(signers[0]), utxoStore, client)
-	if !res.IsOK() {
-		return ctx, res, true
+	if len(signers) != len(spendMsg.Inputs) {
+		return ctx, ErrInvalidSignature(DefaultCodespace, "number of signers does not equal number of signatures").Result(), true
 	}
 
-	// must cover the fee
-	if amt.Cmp(spendMsg.Fee) < 0 {
-		return ctx, ErrInsufficientFee(DefaultCodespace, "first input has an insufficient amount to pay the fee").Result(), true
-	}
-
-	totalInputAmt = amt
-
-	// store confirm signatures
-	if !spendMsg.Input0.Position.IsDeposit() && spendMsg.Input0.TxIndex < 1<<16-1 {
-		plasmaStore.StoreConfirmSignatures(ctx, spendMsg.Input0.Position, spendMsg.Input0.ConfirmSignatures)
-	}
-
-	/* validate second input if applicable */
-	if spendMsg.HasSecondInput() {
-		if len(signers) != 2 {
-			return ctx, msgs.ErrInvalidTransaction(DefaultCodespace, "second signature not present").Result(), true
-		}
-		amt, res = validateInput(ctx, spendMsg.Input1, common.BytesToAddress(signers[1]), utxoStore, client)
+	/* validate inputs */
+	for i, signer := range signers {
```

Comment:
So as I understand it, this is to enable multiple inputs/outputs? Will still need to enforce 2 I/O to be compatible with rootchain right? Where will that check go?

Reply:
transaction.go in plasma will panic when running TxToList if more than 2 I/O are used

Reply:
Does the consensus machine ever run that line tho? I still think there should be an explicit check somewhere during the ValidateBasic process?

Reply:
Also where in TxToList does it panic?

Reply:
I'm not sure it does in TxToList after looking over the code, but it definitely panics somewhere. I tried adding tests cases with 3 inputs and it crashed on me. It seems like we will always have to limits inputs/outputs to some value. Should be hard code a length into ValidateBasic()?

Reply:
the node shouldn't crash. We should just return an error msg